src/Backups/RestoreCoordinationDistributed.cpp

```diff
@@ -0,0 +1,148 @@
+            return;
+        }
+    }
+
+    auto zookeeper = get_zookeeper();
+    String combined_path = zookeeper_path + "/paths_in_backup/" + escapeForFileName(zk_path_);
+    auto code = zookeeper->tryCreate(combined_path, path_in_backup_, zkutil::CreateMode::Persistent);
+    if ((code != Coordination::Error::ZOK) && (code != Coordination::Error::ZNODEEXISTS))
+        throw zkutil::KeeperException(code, combined_path);
+
+    if (code == Coordination::Error::ZNODEEXISTS)
+        path_in_backup_ = zookeeper->get(combined_path);
+
+    {
+        std::lock_guard lock{mutex};
+        paths_in_backup_by_zk_path[zk_path_] = path_in_backup_;
+    }
+}
+
+bool RestoreCoordinationDistributed::acquireZkPathAndName(const String & zk_path_, const String & name_)
+{
+    std::pair<String, String> key{zk_path_, name_};
+
+    {
+        std::lock_guard lock{mutex};
+        if (acquired.contains(key))
+            return true;
+    }
+
+    auto zookeeper = get_zookeeper();
+    String combined_path = zookeeper_path + "/acquired/" + escapeForFileName(zk_path_) + "|" + escapeForFileName(name_);
+    auto code = zookeeper->tryCreate(combined_path, "", zkutil::CreateMode::Persistent);
+    if ((code != Coordination::Error::ZOK) && (code != Coordination::Error::ZNODEEXISTS))
+        throw zkutil::KeeperException(code, combined_path);
+
+    if (code == Coordination::Error::ZNODEEXISTS)
+        return false;
+
+    {
+        std::lock_guard lock{mutex};
+        acquired.emplace(key, std::nullopt);
+        return true;
+    }
+}
+
+void RestoreCoordinationDistributed::setResultForZkPathAndName(const String & zk_path_, const String & name_, Result res_)
+{
+    auto zookeeper = get_zookeeper();
+    String combined_path = zookeeper_path + "/acquired/" + escapeForFileName(zk_path_) + "|" + escapeForFileName(name_);
+    zookeeper->set(combined_path, (res_ == Result::SUCCEEDED) ? "1" : "0");
+
+    {
+        std::lock_guard lock{mutex};
+        acquired[std::pair{zk_path_, name_}] = res_;
+    }
+}
+
+bool RestoreCoordinationDistributed::getResultForZkPathAndName(const String & zk_path_, const String & name_, Result & res_, std::chrono::milliseconds timeout_) const
+{
+    {
+        std::lock_guard lock{mutex};
+        auto value = acquired[std::pair{zk_path_, name_}];
+        if (value)
+        {
+            res_ = *value;
+            return true;
+        }
+    }
+
+    auto zookeeper = get_zookeeper();
+    String combined_path = zookeeper_path + "/acquired/" + escapeForFileName(zk_path_) + "|" + escapeForFileName(name_);
+
+    std::atomic<bool> changed = false;
+    std::condition_variable changed_condvar;
+    const auto watch = [&changed, &changed_condvar, zk_path_, name_](const Coordination::WatchResponse &)
+    {
+        changed = true;
+        changed_condvar.notify_one();
+    };
+
+    String res_str = zookeeper->getWatch(combined_path, nullptr, watch);
+    if (res_str.empty())
+    {
+        std::mutex dummy_mutex;
```

Comment:
Let's use wait for atomics https://en.cppreference.com/w/cpp/atomic/atomic/wait.